pkg/provenance/provenance_test.go

```diff
@@ -0,0 +1,356 @@
+// databases: [moodle wordpress]
+
+func makeSpec(a Args) Spec {
+	attributeToData := make(map[string]interface{}, 0)
+	attributeToData["deploymentName"] = a.DeploymentName
+	attributeToData["image"] = a.Image
+	attributeToData["replicas"] = a.Replicas
+	attributeToData["users"] = a.Users
+	attributeToData["databases"] = a.Databases
+	s := Spec{}
+	s.Version = a.Version
+	s.Timestamp = time.Now().Format("20060102150405")
+	s.AttributeToData = attributeToData
+	return s
+}
+
+// type ObjectLineage map[int]Spec
+
+func initArgs() Args {
+	ds := make([]string, 0)
+	ds = append(ds, "servers")
+	ds = append(ds, "staff")
+
+	users := make([]map[string]string, 0)
+	u1 := make(map[string]string, 0)
+	u1["username"] = "daniel"
+	u1["password"] = "4557832+#^"
+
+	u2 := make(map[string]string, 0)
+	u2["username"] = "kulkarni"
+	u2["password"] = "5*7832!@$"
+
+	users = append(users, u1)
+	users = append(users, u2)
+	a := Args{}
+	a.Users = users
+	a.Databases = ds
+	a.Image = "postgres:9.3"
+	a.Replicas = 3
+	a.Version = 1
+	a.DeploymentName = "MyDeployment"
+
+	return a
+}
+
+// type ObjectLineage map[int]Spec
+// type Spec struct {
+// 	AttributeToData map[string]interface{}
+// 	Version         int
+// 	Timestamp       string
+// }
+//
+// type ProvenanceOfObject struct {
+// 	ObjectFullHistory ObjectLineage
+// 	ResourcePlural    string
+// 	Name              string
+// }
+func deepCopyArgs(arg Args) Args {
+	nArgs := Args{}
+	err := copier.Copy(&nArgs, &arg)
+	if err != nil {
+		fmt.Errorf("Error doing a deep copy of Args: %s", err)
+	}
+
+	//deep copying for maps,
+	// tried two popular libraries and they didn't work..
+	newUsers := make([]map[string]string, 0)
+	for _, m := range nArgs.Users {
+		u := make(map[string]string, 0)
+		for k, v := range m {
+			u[k] = v
+		}
+		newUsers = append(newUsers, u)
+
+	}
+	//The above code successfully made a deep copy of the map
+	nArgs.Users = newUsers
+	return nArgs
+}
+func buildLineage() (ObjectLineage, Args) {
+	objectLineage := make(map[int]Spec)
+	args := initArgs()
+	spec1 := makeSpec(args)
+	objectLineage[spec1.Version] = spec1
+
+	args = deepCopyArgs(args)
+	users2 := addUser(args.Users, "johnson", "fluffy23")
+	args.Users = users2
+	args.Version = 2
+	spec2 := makeSpec(args)
+	objectLineage[spec2.Version] = spec2
+
+	args = deepCopyArgs(args)
+	users3 := addUser(args.Users, "thomas", "bedsheet85")
+	args.Users = users3
+	args.Version = 3
+	spec3 := makeSpec(args)
+	objectLineage[spec3.Version] = spec3
+
+	args = deepCopyArgs(args)
+	db1 := addDatabase(args.Databases, "logging")
+	args.Databases = db1
+	args.Version = 4
+	spec4 := makeSpec(args)
+	objectLineage[spec4.Version] = spec4
+
+	args = deepCopyArgs(args)
+	db2 := addDatabase(args.Databases, "demographics")
+	args.Databases = db2
+	args.Version = 5
+	spec5 := makeSpec(args)
+	objectLineage[spec5.Version] = spec5
+	args = deepCopyArgs(args)
+	return objectLineage, args
+}
+func TestBisectGeneral(t *testing.T) {
```

Comment:
Can you add comment as to what is this test testing?

Looking at the argMapTest data, line 206, and lines 184-191, I understand that database value of demographics will have Version 5. But argMapTest also includes username and password fields. Will they also have Version 5? The initArgs() function  does not assign any version value when setting username 'daniel'.

Reply:
Oh, On line 139 I set the Version of the initial args to be 1 . As you know, all the field/value pairs must be true at a given Version for the bisect to return True
The reason Version 5 should be returned is bc that is when the Database 'demographics' was added, meaning that the query parameters will all be true on Version 5, as the state at that Version contains username daniel, with password 4557832#^, and a database named demographics!